linux/ibus-keyman/src/engine.c

```diff
@@ -150,18 +160,73 @@ ibus_keyman_engine_class_init (IBusKeymanEngineClass *klass)
     engine_class->enable = ibus_keyman_engine_enable;
     engine_class->disable = ibus_keyman_engine_disable;
 
+    engine_class->set_surrounding_text = ibus_keyman_engine_set_surrounding_text;
+    // engine_class->set_cursor_location = ibus_keyman_engine_set_cursor_location;
+
+
     engine_class->focus_in = ibus_keyman_engine_focus_in;
     engine_class->focus_out = ibus_keyman_engine_focus_out;
 
-    engine_class->page_up = ibus_keyman_engine_page_up;
-    engine_class->page_down = ibus_keyman_engine_page_down;
+    // engine_class->page_up = ibus_keyman_engine_page_up;
+    // engine_class->page_down = ibus_keyman_engine_page_down;
 
-    engine_class->cursor_up = ibus_keyman_engine_cursor_up;
-    engine_class->cursor_down = ibus_keyman_engine_cursor_down;
+    // engine_class->cursor_up = ibus_keyman_engine_cursor_up;
+    // engine_class->cursor_down = ibus_keyman_engine_cursor_down;
 
     engine_class->property_activate = ibus_keyman_engine_property_activate;
 }
 
+static gchar *get_current_context_text(km_kbp_context *context)
+{
+    size_t buf_size = 512;
+    km_kbp_context_item *context_items;
+    gchar *current_context_utf8 = g_new0(gchar, buf_size);
+    if (km_kbp_context_get(context, &context_items) == KM_KBP_STATUS_OK) {
+        km_kbp_context_items_to_utf8(context_items,
+                            current_context_utf8,
+                            &buf_size);
+    }
+    km_kbp_context_items_dispose(context_items);
+    g_message("current context is:%lu:%lu:%s:", km_kbp_context_length(context), buf_size, current_context_utf8);
+    return current_context_utf8;
+}
+
+static void reset_context(IBusEngine *engine)
+{
+    IBusKeymanEngine *keyman = (IBusKeymanEngine *) engine;
+    IBusText *text;
+    gchar *surrounding_text, *current_context_utf8;
+    guint cursor_pos, anchor_pos, context_start;
+    km_kbp_context_item *context_items;
+    km_kbp_context *context;
+
+    g_message("reset_context");
+    keyman->firstsurrogate = 0;
+    keyman->marker = False;
+    if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0)
```

Comment:
In the `else` case, do you need to `km_kbp_context_reset()` or equiv?

Reply:
ah, yes, fixed